### ISSUE TITLE
PHP and nginx  is not working

### ISSUE THREAD

**Issue by heyanlong** [Microsoft/BashOnWindows#393]
## php-fpm

ERROR: unable to bind listening socket for address 127.0.0.1:9000 Invalid argument
## Nginx

bind() to 0.0.0.0:81 failed (22: Invalid argument)

by windows 10 preview 14332.rs1_release.160422

**Comment by sunilmut** [comment]
#68

**Comment by sunilmut** [comment]
This is related to #68. I took a deeper look at 'Nginx'. It first binds to an IPv4 address and then to an IPv6 address (to the same port by default) using the IPV6_V6ONLY option. Unfortunately, we are not able to support the scenario immediately and we have a bug tracking this.
As an alternative, there are couple of options as workarounds:
Edit the default configuration of Nginx to start at different ports on IPv4 and IPv6. Or, use either IPv4 or IPv6. The port configuration for Nginx is stored in '/etc/nginx/sites-enabled/default'. I was able to verify that both these workarounds work and the Nginx server starts successfully on build 14342. We know its not ideal, but at least gets your one step further.
Meanwhile, we have an understanding of the bug on our side, but unfortunately, it might take some time to fix it because of dependencies outside of WSL.

**Comment by heyanlong** [comment]
Nginx server starts successfully on build 14342.
But the process will not start work

error.log
```shell
2016/05/20 03:18:24 [alert] 98#0: ioctl(FIOASYNC) failed while spawning "worker process" (22: Invalid argument)
2016/05/20 03:18:24 [alert] 98#0: ioctl(FIOASYNC) failed while spawning "worker process" (22: Invalid argument)
2016/05/20 03:18:24 [alert] 98#0: ioctl(FIOASYNC) failed while spawning "worker process" (22: Invalid argument)
2016/05/20 03:18:24 [alert] 98#0: ioctl(FIOASYNC) failed while spawning "worker process" (22: Invalid argument)
```

**Comment by cjgreen** [comment]
As mentioned in the following blog post [here](https://www.svennd.be/running-nginx-on-bash-for-windows-10/):

Add the following line to /etc/nginx/nginx.conf:
`master_process off;`

Then restart nginx:
`sudo service nginx restart;`

**Comment by williamn** [comment]
I follow the steps taken as the blog post describe and it works. My nginx is now working.
But I have not successful on FPM

I got this when try to start the PHP FPM

`ERROR: failed to retrieve TCP_INFO for socket: Invalid argument`

Any idea why?

**Comment by sunilmut** [comment]
TCP_INFO socket option is currently not supported in WSL, and something we will look into adding. Thanks for the feedback and trying out WSL!

**Comment by PeterDraex** [comment]
How can I prevent FPM from using TCP_INFO socket option? Thanks. @sunilmut @williamn

**Comment by loutian** [comment]
@cjgreen   I got the same error,  did you solved it ?

**Comment by spagu** [comment]
any solution so far?

**Comment by heyanlong** [comment]
@spagu i'm not retry

**Comment by stehufntdev** [comment]
After adding signal driven IO support for unix stream sockets, I was able to run nginx locally and load the default site.  Also confirmed that the errors were gone from the nginx error log.  The fix should be out in insider builds soon.

Can you please confirm if the issues with php are resolved?

**Comment by chrisboylan** [comment]
Just updated to 15019 and the PHP problem still persists. Do you have a rough idea of when/which build it will be in?

**Comment by stehufntdev** [comment]
Thanks for checking back.  The fix didn't make it into the last insider build, but it should be in soon.  When that happens we'll include it in the release notes and mark this thread as fixedinsider.

**Comment by chrisboylan** [comment]
Semi-success! Upgraded to 15025 and can get php7.0 and nginx to run. phpinfo(); works, but Im having a tough time getting any apps to run. Getting a lot of session writing errors:

`FastCGI sent in stderr: "PHP message: PHP Warning:  session_write_close(): Failed to write session data (user). Please verify that the current setting of session.save_path is correct (/var/lib/php/sessions)`

Tried changing permissions on /var/lib/php/sessions to no avail. Going to try and point it elsewhere.

**Comment by stehufntdev** [comment]
Thanks for the update, can you please provide a strace of the failure and repro steps of what you tried?

**Comment by chrisboylan** [comment]
Funny thing is that I was checking phpinfo() for details, and it's getting cut off with a really fast timeout that I can't figure out. 

The error.log line from nginx is 

`2017/02/03 01:33:31 [error] 1158#1158: *1 upstream timed out (110: Connection timed out) while reading upstream, client: 127.0.0.1, server: xxxxxx.com, request: "GET /phpinfo.php HTTP/1.1", upstream: "fastcgi://unix:/var/run/php/php7.0-fpm.sock:", host: "xxxxxx.loc"`

The nginx.conf line for PHP on this site is:

` location ~ \.php$ {
    try_files $uri =404;
    fastcgi_split_path_info ^(.+\.php)(/.+)$;
    fastcgi_pass unix:/var/run/php/php7.0-fpm.sock;
    fastcgi_index index.php;
    fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
    include fastcgi_params;
    fastcgi_read_timeout 120; #shot in dark to end timeout problems (didn't work)
  }`

It's entirely possible its just a bad config on my side somehow. Think I might nuke the WSL install and try again.

**Comment by stehufntdev** [comment]
Thanks, please let us know if you hit any other issues.  If you do, a strace and repro steps will help us to narrow down the cause.

**Comment by lucas-sandery** [comment]
If, like me, you're not on the insider builds but don't want your logs filling up with this:

    ERROR: failed to retrieve TCP_INFO for socket: Invalid argument (22)

You can set `log_level = alert` in `/etc/php5/fpm/php-fpm.conf`. Obviously not ideal but better than thrashing your SSD.

**Comment by dvaknheo** [comment]
I take the same problem as 
chrisboylan commented on 3 Feb
and have no solution;

**Comment by cute-angelia** [comment]
I take the same problem

but the port 9000 is on ~

**Comment by dvaknheo** [comment]
solute it , do not run  nginx/php7.0-fpm by manual in command line.
just 
sudo service php-7.0-fpm start
sudo service nginx start

**Comment by dvaknheo** [comment]
err , i am wrong ,
you must use port than unix file socket , otherwise you php  must be run slowlest

**Comment by DarthSpock** [comment]
@dvaknheo when you say port rather than unix file socket, are you referring to `SOCK_STREAM` that was brought to Windows in build 17063? or the native file socket in WSL? Per this [blog](https://blogs.msdn.microsoft.com/commandline/2017/12/19/af_unix-comes-to-windows/), support to WSL should be coming based on this "And, if you are wondering, there is already support for unix socket within Windows Subsystem for Linux (WSL), how does that work with the Windows unix socket implementation? Well, currently, it doesn’t, but stay tuned! "

**Comment by sunilmut** [comment]
@DarthSpock - What is the problem that you are facing? Can you please post the repro steps and the output of the following command from the bash prompt `cmd.exe /c ver`?

The `AF_UNIX` blog post you are referring to should not have anything to do with the functioning of `PHP` or `nginx` within WSL.

**Comment by DarthSpock** [comment]
@sunilmut I wasn't providing an issue. I'm just curious if @dvaknheo has tested performance with the latest Insider build as I'm not able to install Insider builds. I don't normally work with nginx but do play slightly with php. I brought the article up as a case of eventually having a possible performance boost for utilizing Unix sockets vice tcp ports which is currently done in some situations such as using nginx

**Comment by sunilmut** [comment]
@DarthSpock - Got it, thanks! A forward looking statement, but theoretically speaking (without any measurements), the way things are currently implemented and layered, I would not expect a difference in performance between WSL<->Win32 between TCP loopback and AF_UNIX sockets.  Note that this is just in theory.

**Comment by YouHeng1** [comment]
In nginx error log there is such an error message.
`2018/01/21 00:29:22 [emerg] 35#35: io_setup() failed (38: Function not implemented)`

**Comment by therealkenc** [comment]
> 2018/01/21 00:29:22 [emerg] 35#35: io_setup() failed (38: Function not implemented)

Open a new one for that following CONTRIBUTING.md. Cant tell you if it is a dupe or not because not enough information provided. This issue (`bind() to 0.0.0.0:81 failed (22: Invalid argument)` was closed a long time ago along with #68.

**Comment by sunilmut** [comment]
@YouHeng1 - Yes, please open a new issue for this and also note any functionality issue that you might be observing with nginx.

**Comment by heyanlong** [comment]
@dvaknheo 
多写代码，少玩游戏
Write more code, play less games

**Comment by hparadiz** [comment]
This configuration works in Ubuntu WSL.
Please note that `fastcgi_buffering` is `off`.
Without this nginx will wait on PHP forever and never complete.
```nginx
location ~ \.php$ {
        proxy_set_header X-Forwarded-Proto $scheme;
        fastcgi_split_path_info ^(.+\.php)(/.+)$;
        fastcgi_pass unix:/var/run/php/php7.1-fpm.sock;
        fastcgi_index index.php;
        include fastcgi_params;
        fastcgi_buffering off; # This must be here for WSL as of 11/28/2018
        fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
        fastcgi_param PHP_VALUE "upload_max_filesize = 20M \n post_max_size=21M";
        include /etc/nginx/fastcgi.conf;
}
```

**Comment by Babaktrad** [comment]
> This configuration works in Ubuntu WSL.
> Please note that `fastcgi_buffering` is `off`.
> Without this nginx will wait on PHP forever and never complete.
> 
> ```nginx
> location ~ \.php$ {
>         proxy_set_header X-Forwarded-Proto $scheme;
>         fastcgi_split_path_info ^(.+\.php)(/.+)$;
>         fastcgi_pass unix:/var/run/php/php7.1-fpm.sock;
>         fastcgi_index index.php;
>         include fastcgi_params;
>         fastcgi_buffering off; # This must be here for WSL as of 11/28/2018
>         fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
> }
> ```>         fastcgi_param PHP_VALUE "upload_max_filesize = 20M \n post_max_size=21M";
>         include /etc/nginx/fastcgi.conf;


It works for me. It was nearly two weeks that I was involved with this issue! Thanks a lot.

**Comment by PhillR** [comment]
@hparadiz Hero answer, solved a long, outstanding problem for me. Thank you!

**Comment by phpcat100** [comment]
thank you.

**Comment by tawhidulIKhan** [comment]
It works! thanks @hparadiz
